### PR TITLE
Add more encoder tests and polish APIs

### PR DESCRIPTION
**Pull request by tmat**

### CODE REVIEW

**Comment by tmat** [comment]
@nguerrera

**Comment by tmat** [comment]
test this please

**Comment by nguerrera** [comment]
:+1:

**Comment by tmat** [comment]
test this please

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs [hunk 1]
```diff
@@ -1114,11 +1114,11 @@ public void WriteCompressedSignedInteger(int value)
         /// 
         /// Otherwise, encode as a 4-byte integer, with bit 31 set, bit 30 set, bit 29 clear (value held in bits 28 through 0).
         /// </remarks>
-        /// <exception cref="ArgumentOutOfRangeException"><paramref name="value"/> can't be represented as a compressed integer.</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="value"/> can't be represented as a compressed unsigned integer.</exception>
         /// <exception cref="InvalidOperationException">Builder is not writable, it has been linked with another one.</exception>
         public void WriteCompressedInteger(int value)
         {
-            BlobWriterImpl.WriteCompressedInteger(this, value);
+            BlobWriterImpl.WriteCompressedInteger(this, unchecked((uint)value));
```
Comment (stephentoub) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs]: Do we have checked enabled by default for this project?

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/BlobEncoders.cs [hunk 2]
```diff
@@ -930,16 +980,54 @@ public ArrayShapeEncoder(BlobBuilder builder)
...
+
             Builder.WriteCompressedInteger(sizes.Length);
             foreach (int size in sizes)
             {
                 Builder.WriteCompressedInteger(size);
             }
 
-            if (lowerBounds.IsDefault)
+            // lower bounds
+            if (lowerBounds.IsDefault) // TODO: remove -- update Roslyn
```
Comment (stephentoub) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/BlobEncoders.cs]: Is there an issue tracking TODOs like this?

**Comment by stephentoub** [comment]
Consider updating the commit description, as this is not just tests.

**Comment by stephentoub** [comment]
LGTM

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/BlobEncoders.cs (hunk 2 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/Ecma335/Encoding/BlobEncoders.cs]: I'm tracking it in my list of work to do on SRM this sprint to reduce overhead. Will convert any TODO's remaining after the sprint to git issues.

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs (hunk 1 above)
Comment (tmat) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/BlobBuilder.cs]: Regardless of the setting I'm using unchecked here to express the intent that the overflow is expected.

**Comment by tmat** [comment]
test this please

**Pull request merged**